stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+inline return_type_t<T_omega, T_Gamma, T_rho> hmm_marginal_lpdf(
+    const Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>& log_omegas,
+    const Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>& Gamma,
+    const Eigen::Matrix<T_rho, Eigen::Dynamic, 1>& rho) {
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  check_square("hmm_marginal_lpdf", "Gamma", Gamma);
+  check_consistent_size("hmm_marginal_lpdf", "Gamma", row(Gamma, 1), n_states);
+  {
+    // Temporary vector to use check_simplex, which only works once
+    // column vectors.
+    Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic> Gamma_transpose
+        = Gamma.transpose();
+    for (int i = 0; i < Gamma.cols(); i++) {
+      // CHECK -- does check_simplex not work on row-vectors?
+      check_simplex("hmm_marginal_lpdf", "Gamma[i, ]",
+                    col(Gamma_transpose, i + 1));
+    }
+  }
+  check_consistent_size("hmm_marginal_lpdf", "rho", rho, n_states);
+  check_simplex("hmm_marginal_lpdf", "rho", rho);
+
+  using T_partials_return = partials_return_t<T_omega, T_Gamma, T_rho>;
+  operands_and_partials<Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_rho, Eigen::Dynamic, 1> >
+      ops_partials(log_omegas, Gamma, rho);
+
+  Eigen::MatrixXd alphas(n_states, n_transitions + 1);
+  Eigen::VectorXd alpha_log_norms(n_transitions + 1);
+  Eigen::MatrixXd omegas;
+  Eigen::MatrixXd Gamma_dbl = value_of_rec(Gamma);
+
+  T_partials_return log_marginal_density
+      = hmm_marginal_lpdf(value_of_rec(log_omegas), Gamma_dbl,
+                          value_of_rec(rho), alphas, alpha_log_norms, omegas);
+
+  // Variables required for all three Jacobian-adjoint products.
+  double norm_norm = alpha_log_norms(n_transitions);
+  double unnormed_marginal = alphas.col(n_transitions).sum();
+
+  std::vector<Eigen::VectorXd> kappa(n_transitions);
+  Eigen::VectorXd kappa_log_norms(n_transitions);
+  std::vector<double> grad_corr(n_transitions);
+
+  if (n_transitions > 0) {
+    kappa[n_transitions - 1] = Eigen::VectorXd::Ones(n_states);
+    kappa_log_norms(n_transitions - 1) = 0;
+    grad_corr[n_transitions - 1]
+        = std::exp(alpha_log_norms(n_transitions - 1) - norm_norm);
+  }
+
+  for (int n = n_transitions - 2; n >= 0; --n) {
+    kappa[n] = Gamma_dbl * (omegas.col(n + 2).cwiseProduct(kappa[n + 1]));
+
+    double norm = kappa[n].maxCoeff();
+    kappa[n] /= norm;
+    kappa_log_norms(n) = std::log(norm) + kappa_log_norms(n + 1);
+    grad_corr[n]
+        = std::exp(alpha_log_norms(n) + kappa_log_norms(n) - norm_norm);
+  }
+
+  if (!is_constant_all<T_Gamma>::value) {
+    Eigen::MatrixXd Gamma_jacad(n_states, n_states);
+    Gamma_jacad.setZero();
+
+    if (n_transitions != 0) {
+      for (int n = n_transitions - 1; n >= 0; --n) {
```

Comment:
The idiom here is the following:

```
for (int n = n_transitions; n_transitions-- > 0; )
```

Mainly because it'll work if the type is an unsized type, but also becaue it reduces the number of operations.

Reply:
I don't follow. The given code doesn't work (and nor do small tweaks around it). Can you be more specific about the wanted change?